packages/payment-widget/src/lib/utils/request.ts

```diff
@@ -0,0 +1,238 @@
+import type { Currency } from "../types";
+import { Contract, providers, utils } from "ethers";
+import {
+  hasSufficientFunds,
+  approveErc20,
+  payRequest,
+} from "@requestnetwork/payment-processor";
+import { chains } from "./chains";
+
+export const prepareRequestParameters = ({
+  currency,
+  sellerAddress,
+  payerAddress,
+  amountInCrypto,
+  exchangeRate,
+  amountInUSD,
+}: {
+  currency: Currency;
+  sellerAddress: string;
+  payerAddress: string;
+  amountInCrypto: number;
+  exchangeRate: number;
+  amountInUSD: number;
+}) => {
+  const isERC20 = currency.type === Types.RequestLogic.CURRENCY.ERC20;
+  const currencyValue = isERC20 ? currency.address : "eth";
+
+  return {
+    requestInfo: {
+      currency: {
+        type: currency.type,
+        value: currencyValue,
+        network: currency.network,
+      },
+      expectedAmount: utils
+        .parseUnits(amountInCrypto.toString(), currency.decimals)
+        .toString(),
+      payee: {
+        type: Types.Identity.TYPE.ETHEREUM_ADDRESS,
+        value: sellerAddress,
+      },
+      payer: {
+        type: Types.Identity.TYPE.ETHEREUM_ADDRESS,
+        value: payerAddress,
+      },
+      timestamp: Utils.getCurrentTimestampInSecond(),
+    },
+    paymentNetwork: {
+      id: isERC20
+        ? Types.Extension.PAYMENT_NETWORK_ID.ERC20_FEE_PROXY_CONTRACT
+        : Types.Extension.PAYMENT_NETWORK_ID.ETH_FEE_PROXY_CONTRACT,
+      parameters: {
+        paymentNetworkName: currency.network,
+        paymentAddress: sellerAddress,
+        feeAddress: "0x0000000000000000000000000000000000000000",
+        feeAmount: "0",
+        tokenAddress: currencyValue,
+      },
+    },
+    contentData: {
+      paymentCurrency: {
+        type: currency.type,
+        value: currencyValue,
+        network: currency.network,
+      },
+      exchangeRate: exchangeRate.toString(),
+      amountInUSD: amountInUSD.toString(),
+    },
+    signer: {
+      type: Types.Identity.TYPE.ETHEREUM_ADDRESS,
+      value: payerAddress,
+    },
+  };
+};
+
+export const handleRequestPayment = async ({
+  requestParameters,
+  walletProvider,
+  payerAddress,
+  persistRequest,
+}: {
+  requestParameters: any;
+  walletProvider: any;
+  payerAddress: string;
+  persistRequest: boolean;
+}) => {
+  let ethersProvider: providers.Web3Provider;
+  let targetChain: (typeof chains)[0];
+
+  const initializeProvider = async () => {
+    ethersProvider = new providers.Web3Provider(walletProvider);
+    const targetNetwork = requestParameters.requestInfo.currency.network;
+    const chain = getChainFromNetwork(targetNetwork);
+    if (!chain) {
+      throw new Error(`Unsupported network: ${targetNetwork}`);
+    }
+    targetChain = chain;
+  };
+
+  const ensureCorrectNetwork = async () => {
+    const currentChainId = await ethersProvider
+      .getNetwork()
+      .then((net) => net.chainId);
+
+    if (currentChainId !== targetChain.chainId) {
+      try {
```

Comment:
Do we need two try catch blocks? Can we just push the error higher and catch it there?

Reply:
Different type of errors being throw. this allows for specificity to which error message to deal with.